design/proposals/variable_schema_validation.md

```diff
@@ -0,0 +1,264 @@
+      "type": "integer",
+      "minimum": 1,
+      "maximum": 65535
+    }
+  },
+  "required": ["username", "password"],
+  "additionalProperties": false,
+  "secrets": [
+      "password",
+    ]
+}
+```
+
+And the `mysql-db-creds` variable could be updated with the following value:
+
+```json
+{
+  "username": "my-user",
+  "password": "my-password-12345",
+  "address": "my-db.com",
+  "port": 3306
+}
+```
+
+### Functionality
+
+#### Input Validation
+
+Conjur will validate the variable content, using the JSON schema that is specified in the `schema` attribure of that variable.
+
+For the example above, Conjur will use the JSON schema specified in `conjur/schemas/mysql-schema` to verify the content of `mysql-db-creds`. This verification includes:
+
+- `username` and `password` are mandatory properties.
+- `address` and `port` are allowed properties.
+- `port` must contain a valid port number.
+- `password` must align to a sufficient complexity.
+
+If the varialbe content is invalid, the user will get an HTTP code of 422 (Unprocessable Entity) and the body of the response will also contain an elaborative error message that explains what part of the input was found to be invalid.
+
+#### Secrets Encryption
+
+The JSON schema can contain a property called `secrets` which expects an array of JSON schema property names. The properties specified in this array, will be encrypted in the database. All the rest of the properties and the JSON structure will remain in cleartext. This allows the protection of sensitive data while also allowing the non-sensitive values to be searchable.
+
+For the example above, the `mysql-db-creds` variable will be saved in the database as follows:
+
+```json
+{
+  "username": "my-user",
+  "password": "U2FsdGVkX19Ji6JpgnVHW3V47OtMJwuKi1Yf9nc0aP5QcuzdnIrpzZ2zMC90f24g",
+  "address": "my-db.com",
+  "port": 3306
+}
+```
+
+Before the variable value is returned to the client, the secret attributes are decrypted so that the given example above, would look like this:
+
+```json
+{
+  "username": "my-user",
+  "password": "my-password",
+  "address": "my-db.com",
+  "port": 3306
+}
+```
+
+#### Builtin Schemas
+
+To simplify the user experience, Conjur can come with prdefined JSON schemas for the most common secrets use cases. The list of predefined schmeas should include:
+
+- Database secrets - Oracle, MSSQL, MySQL, Postgres, MariaDB, DB2.
+- Cloud access keys - AWS, Azure, GCP.
+- X509 certificates.
+- JWTs
+
+Write about the advantages of atomicity, minimal changes in the database to implement this and so on.
+
+#### Additional Required Functionality
+
+In order to provide access for any role to the builtin schemas, a new builtin group should be introduced: `conjur/all`. All roles in the Conjur account should be added to this group automatically, thus keeping this group always up to date with all the roles that exist in the account.
+
+Usage example:
+
+```yaml
+- !permit
+  role: !group /conjur/all
```

Comment:
I really like this idea! Is this part of the core design? As in can we go forth with the design without implementing this or is it an enhancement?

Reply:
This isn't a minimal requirement, because the admin could manage access to the builtin schema, but for a better user experience, it would be useful to just have access to them automatically

Reply:
I think this idea is very good because it betters user experience